Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -719,6 +723,61 @@ ObjectValue::Map DecodeObject(Reader* reader) {
       });
 }
 
+/**
+ * Creates the prefix for a fully qualified resource path, without a local path
+ * on the end.
+ */
+ResourcePath EncodeDatabaseId(const DatabaseId& database_id) {
+  return ResourcePath{"projects", database_id.project_id(), "databases",
+                      database_id.database_id()};
+}
+
+/**
+ * Encodes a databaseId and resource path into the following form:
+ * /projects/$projectId/database/$databaseId/documents/$path
+ */
+std::string EncodeResourceName(const DatabaseId& database_id,
+                               const ResourcePath& path) {
+  return EncodeDatabaseId(database_id)
+      .Append("documents")
+      .Append(path)
+      .CanonicalString();
+}
+
+/** Validates that a path has a prefix that looks like a valid encoded
+ * databaseId. */
+bool IsValidResourceName(const ResourcePath& path) {
+  // Resource names have at least 4 components (project ID, database ID)
+  // and commonly the (root) resource type, e.g. documents
```

Comment:
Optional nit: since I don't think this code checks this part "and commonly the (root) resource type, e.g. documents", perhaps omit it? The previous part already mentions that there might be more components.

Reply:
Ah, thanks for explaining. In that case, it's likely not worth the trouble.